git/dependencies.py

```diff
@@ -0,0 +1,415 @@
+					print('Error: unknown ref: ' + ref)
+					return
+
+		self.config[p]['ref'] = ref
+		self.__saveDependenciesFile()
+
+	def freezeDependency(self, path = '*', recursive = False):
+		if (self.config == None):
+			return
+
+		cleanPath = self.__cleanPath(path)
+		if (cleanPath == '*'):
+			sections = self.config.sections()
+		else:
+			sections = [cleanPath]
+		for p in sections:
+			if 'freezed' in self.config[p]:
+				continue
+
+			dependencyPath = os.path.join(self.repositoryPath, p)
+			if (not os.path.exists(dependencyPath)):
+				self.updateDependencies(p, False)
+
+			d = GitDependenciesRepository(self.config[p]['url'], dependencyPath)
+			if (recursive):
+				d.freezeDependency('*', recursive)
+
+			hash = d.currentSha()
+			self.config[p]['freezed'] = self.config[p]['ref']
+			self.config[p]['ref'] = hash
+
+			self.updateDependencies(p, False)
+		self.__saveDependenciesFile()
+		self.commit(message = 'Freezing dependencies', files = [GITDEPENDS_FILE])
+
+	def unfreezeDependency(self, path = '*', recursive = False):
+		if (self.config == None):
+			return
+
+		cleanPath = self.__cleanPath(path)
+		if (cleanPath == '*'):
+			sections = self.config.sections()
+		else:
+			sections = [cleanPath]
+		for p in sections:
+			if 'freezed' not in self.config[p]:
+				continue
+
+			self.config[p]['ref'] = self.config[p]['freezed']
+			self.config.remove_option(p, 'freezed')
+
+			self.updateDependencies(p, False)
+
+			d = GitDependenciesRepository(self.config[p]['url'], os.path.join(self.repositoryPath, p))
+			if (recursive):
+				d.unfreezeDependency('*', recursive)
+
+			# self.checkout(self.config[p]['ref'])
+		self.__saveDependenciesFile()
+		self.commit(message = 'Unfreezing dependencies', files = [GITDEPENDS_FILE])
+
+	def foreachDependency(self, command, recursive = False):
+		if (self.config == None):
+			return
+
+		runCommand = command
+		if (runCommand.startswith('!sh ')):
+			runCommand = command[4:]
+			i = runCommand.index(' ')
+			task = Task(runCommand[:i])
+			runCommand = runCommand[i + 1:]
+		else:
+			task = Task('git')
+
+		for p in self.config.sections():
+			dependencyPath = os.path.join(self.repositoryPath, p)
+			if (self.__isSymlink(dependencyPath)):
+				realPath = self.__resolveSymlinkRealPath(dependencyPath)
+				print(p + " is a symlink to " + realPath + ", skipping from foreach.")
+				continue
+
+			d = GitDependenciesRepository(self.config[p]['url'], dependencyPath)
+			wd = os.getcwd()
+			os.chdir(d.repositoryPath)
+			task.run(runCommand.split(' '))
+			if (task.output != ''):
+				print(task.output, file=sys.stderr)
+			if (task.exitCode() != 0):
+				sys.exit(task.exitCode())
+			os.chdir(wd)
+			if (recursive):
+				d.foreachDependency(command, True)
+
+	def command(self, path, command):
+		if (self.config == None):
+			return
+
+		p = self.__cleanPath(path)
+
+		if command and len(command) > 0:
+			self.config[p]['command'] = "{}".format(command)
+		else:
+			self.config.remove_option(p, 'command')
```

Comment:
This command option will be removed if the given command is empty. In my view it is unnecessary to store command option without value.